provider/oracle/storage_volumes.go

```diff
@@ -0,0 +1,558 @@
+		return false, errors.Trace(err)
+	}
+	return string(details.State) == desiredStatus, nil
+}
+
+// waitForResourceStatus will ping the resource until the fetch function returns true,
+// the timeout is reached, or an error occurs.
+func (o *oracleVolumeSource) waitForResourceStatus(
+	fetch func(name string, desiredStatus string) (complete bool, err error),
+	name, state string, timeout time.Duration) error {
+	errChan := make(chan error)
+	done := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				done, err := fetch(name, state)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if done {
+					errChan <- nil
+					return
+				}
+				time.Sleep(2 * time.Second)
+			}
+		}
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-o.clock.After(timeout):
+		done <- true
+		return errors.Errorf(
+			"timed out waiting for resource %q to transition to %v",
+			name, state,
+		)
+	}
+	return nil
+}
+
+// ListVolumes is specified on the storage.VolumeSource interface.
+func (s *oracleVolumeSource) ListVolumes() ([]string, error) {
+	tag := fmt.Sprintf("%s=%s", tags.JujuModel, s.modelUUID)
+	filter := []oci.Filter{
+		oci.Filter{
+			Arg:   "tags",
+			Value: tag,
+		},
+	}
+	volumes, err := s.api.AllStorageVolumes(filter)
+	if err != nil {
+		return nil, errors.Annotate(err, "listing volumes")
+	}
+
+	ids := make([]string, len(volumes.Result))
+	for i, volume := range volumes.Result {
+		ids[i] = volume.Name
+	}
+
+	return ids, nil
+}
+
+// DescribeVolumes is specified on the storage.VolumeSource interface.
+func (s *oracleVolumeSource) DescribeVolumes(volIds []string) ([]storage.DescribeVolumesResult, error) {
+	if volIds == nil || len(volIds) == 0 {
+		return []storage.DescribeVolumesResult{}, nil
+	}
+
+	tag := fmt.Sprintf("%s=%s", tags.JujuModel, s.modelUUID)
+	filter := []oci.Filter{
+		oci.Filter{
+			Arg:   "tags",
+			Value: tag,
+		},
+	}
+
+	result := make([]storage.DescribeVolumesResult, len(volIds), len(volIds))
+	volumes, err := s.api.AllStorageVolumes(filter)
+	if err != nil {
+		return nil, errors.Annotatef(err, "describe volumes")
+	}
+	asMap := map[string]ociResponse.StorageVolume{}
+	for _, val := range volumes.Result {
+		asMap[val.Name] = val
+	}
+	for i, volume := range volIds {
+		if vol, ok := asMap[volume]; ok {
+			volumeInfo := &storage.VolumeInfo{
+				VolumeId:   vol.Name,
+				Size:       uint64(vol.Size) / 1024 / 1024 / 1024,
+				Persistent: true,
+			}
+			v := storage.DescribeVolumesResult{
+				VolumeInfo: volumeInfo,
+			}
+			result[i] = v
+		} else {
+			result[i].Error = errors.NotFoundf("%s", volume)
+		}
+	}
+	return result, nil
+}
+
+// DestroyVolumes is specified on the storage.VolumeSource interface.
+func (s *oracleVolumeSource) DestroyVolumes(volIds []string) ([]error, error) {
+	results := make([]error, len(volIds))
+	wg := sync.WaitGroup{}
+	wg.Add(len(volIds))
+	for i, val := range volIds {
+		go func(volId string) {
```

Comment:
you must pass `i` in as well; the value of the index will be overwritten with each iteration, and all goroutines will end up referring to the same index